velodyne_pointcloud/include/velodyne_pointcloud/datacontainerbase.h

```diff
@@ -181,13 +174,39 @@ class DataContainerBase
     Eigen::Vector3f eigen_origin;
     vectorTfToEigen(transform.getOrigin(), eigen_origin);
     Eigen::Translation3f translation(eigen_origin);
-    transformation = translation * rotation;
+    htm = translation * rotation;
     return true;
   }
 
+  inline bool computeTransformToTarget(const ros::Time &scan_time)
+  {
+    std::string &source_frame = config_.fixed_frame.empty()
+                                    ? cloud.header.frame_id
+                                    : config_.fixed_frame;
+    return !(!config_.target_frame.empty() &&
+             !lookupHTM(htm_to_target, config_.target_frame, source_frame,
+                        scan_time));
+  }
+
+  inline bool computeTransformToFixed(const ros::Time &packet_time)
+  {
+    std::string &source_frame = cloud.header.frame_id;
+    return !(!config_.fixed_frame.empty() &&
+             !lookupHTM(htm_to_fixed, config_.fixed_frame, source_frame,
+                        packet_time));
+  }
+
```

Comment:
Can you explain why we need two transformations here. The `fixed_frame` should be defined by the user as the static coordinate frame, i.e. the root of the tf tree. The `target_frame` is the goal coordinate frame to which we want to transform the points, this can of course also be configured to be the `fixed_frame`. The `source_frame` should be the `sensor_frame` which is for example defined in the URDF. Another thing are the two options 1.) transform packet wise or 2.) transform once per cloud. Am I right? Please explain a bit here, if I am wrong. What am I missing here?

Reply:
I mean you are separating the transformation into two parts `fixed_frame`->`target_frame`, or `cloud.header.frame_id`->`target_frame`, where the `cloud.header.frame_id` is set either to the `target_frame` in line 121 or to `fixed_frame` in line 125 and the second transformation `source_frame` -> `fixed_frame'  where the `source_frame` is again the `cloud.header.frame_id`, so line 121 or 125. Could you maybe explain this and make some comments in the code or a simple sketch. Thank you.

Reply:
Maybe it helps to read through issue #342  (including watching the video) if you haven't already.

The idea is that the point cloud has to be transformed packet-wise (actually even better point-wise based on its timestamp, however this would be extremely CPU intensive) into the odom frame based on the packet timestamp. At this point one automatically obtains a ego motion compensated point cloud where the point coordinates are specified w.r.t. odom. This is what `computeTransformToFixed` is intended for. When fixed_frame is not empty it computes `htm_to_fixed` for subsequent use. This matrix transforms points from sensor coords (already stored in `cloud.header.frame_id` beforehand, compare line 104) to the fixed frame.

Points in odom coords are sometimes undesired, especially when someone wants to perform downstream segmentation tasks. In this case one typically wants the points to be back in sensor coordinates, however this time ego motion compensated. Transforming the whole point cloud given a single timestamp (single transformation matrix for whole point cloud) achieves this. This point cloud is different to directly converting to target frame:

![notes](https://user-images.githubusercontent.com/11663218/85701696-f1dc5d80-b6dd-11ea-894d-321801779865.jpg)

This is what `computeTransformToTarget` is intended for. Here the source frame is here selected based on the fact whether ego motion compensation was performed (odom) or not (points are still in sensor frame, stored inside `cloud.header.frame_id` in line 104). Lines 121 and 125 are called after all these transformation when the point cloud is finished, so they have no effect here. You are right when you say that it is a little bit confusing that the string of the sensor frame is stored/passed in the point cloud message and not explicitly passed in a dedicated parameter/variable. However, this is a thing that was already implemented this way before my pull request and you already made clear that I should not perform any refactoring in this pull request.